src/ApplicationInsights.ServiceFabric.Native/Net45/CorrelatingRemotingMessageHandler.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            this.innerHandler = new ActorServiceRemotingDispatcher(actorService);
+            this.serviceContext = actorService.Context;
+            Initialize();
+        }
+
+        /// <summary>
+        /// Handles a one way message from the client. It consumes the correlation id and context from the message headers, if any.
+        /// </summary>
+        /// <param name="requestContext">Request context - contains additional information about the request.</param>
+        /// <param name="messageHeaders">Request message headers.</param>
+        /// <param name="requestBody">Request message body.</param>
+        public void HandleOneWay(IServiceRemotingRequestContext requestContext, ServiceRemotingMessageHeaders messageHeaders, byte[] requestBody)
+        {
+            HandleAndTrackRequestAsync(messageHeaders, () => 
+            {
+                this.innerHandler.HandleOneWay(requestContext, messageHeaders, requestBody);
+                return Task.FromResult<byte[]>(null);
+            }).Forget();
+        }
+
+        /// <summary>
+        /// Handles a message from the client that requires a response from the service. It consumes the correlation id and
+        /// context from the message headers, if any.
+        /// </summary>
+        /// <param name="requestContext">Request context - contains additional information about the request.</param>
+        /// <param name="messageHeaders">Request message headers.</param>
+        /// <param name="requestBody">Request message body</param>
+        /// <returns></returns>
+        public Task<byte[]> RequestResponseAsync(IServiceRemotingRequestContext requestContext, ServiceRemotingMessageHeaders messageHeaders, byte[] requestBody)
+        {
+            return HandleAndTrackRequestAsync(messageHeaders, () => this.innerHandler.RequestResponseAsync(requestContext, messageHeaders, requestBody));
+        }
+
+        private void Initialize()
+        {
+            this.telemetryClient = new TelemetryClient();
+            this.baggageSerializer = new Lazy<DataContractSerializer>(() => new DataContractSerializer(typeof(IEnumerable<KeyValuePair<string, string>>)));
+
+            // TODO: SF should expose method name without the need to use reflection
+            this.methodMap = typeof(ServiceRemotingDispatcher)?.GetField("methodDispatcherMap", BindingFlags.Instance | BindingFlags.NonPublic)
+                                ?.GetValue(this.innerHandler) as IDictionary<int, ServiceMethodDispatcherBase>;
+        }
+
+        private async Task<byte[]> HandleAndTrackRequestAsync(ServiceRemotingMessageHeaders messageHeaders, Func<Task<byte[]>> doHandleRequest)
+        {
+            // Create and prepare activity and RequestTelemetry objects to track this request.
+            RequestTelemetry rt = new RequestTelemetry();
+
+            if (messageHeaders.TryGetHeaderValue(ServiceRemotingLoggingStrings.ParentIdHeaderName, out string parentId) &&
```

Comment:
having 2 headers is a simplification for customers who do not want to deal with hierarchical ids.
But we know how to deal with them. It also creates yet another header schema.
I would still prefer to keep 1 header and if you really badly want to avoid using Activity instead of StartOperation, just parse the root id from the request id.